dandiapi/zarr/views/__init__.py

```diff
@@ -139,7 +140,10 @@ def create(self, request):
             raise ValidationError('Cannot add zarr to embargoed dandiset')
         zarr_archive: ZarrArchive = ZarrArchive(name=name, dandiset=dandiset)
         try:
-            zarr_archive.save()
+            with transaction.atomic():
+                zarr_archive.save()
+
+                audit.create_zarr(dandiset=dandiset, user=request.user, zarr_archive=zarr_archive)
         except IntegrityError as e:
```

Comment:
I'm a little worried about the behavior here, as now there's two separate things that could cause an `IntegrityError` to be raised:

1. The existing behavior, which is that we attempt to create a zarr, and one already exists.
2. Some other `IntegrityError` from the audit record creation


Honestly I'm not even sure why the zarr is being created this way. I think we should simply replace that implementation with the following, and then the atomic block could exist without any exception handling.

```python
if ZarrArchive.objects.filter(name=name, dandiset=dandiset).exists():
    raise ValidationError('Zarr already exists')

with transaction.atomic():
    zarr_archive = ZarrArchive.objects.create(name=name, dandiset=dandiset)
    audit.create_zarr(dandiset=dandiset, user=request.user, zarr_archive=zarr_archive)
```

To prevent race conditions we also might want to do a `select_for_update`. This might be too out of scope for this PR, so a TODO and issue would suffice.

Reply:
In 7e04c83b8d4c0e081f6cc7b0420dd721731ea014 I've attempted to solve both problems using nested transactions (following advice from the "avoid catching exceptions" callout in the [documentation on transactions](https://docs.djangoproject.com/en/5.0/topics/db/transactions/#controlling-transactions-explicitly)).

Let me know what you think.

Reply:
Looks good to me! I made a suggestion to just include a comment above that block to further clarify things.